### PR TITLE
Bug/user info query

### PR DESCRIPTION
**Pull request by MikhailSorokin**
# Pull Request Template

## Checklist:

- [X] The code follows the style guidelines of this project and the [Epic Coding Standard](https://docs.unrealengine.com/en-US/Programming/Development/CodingStandard/index.html)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have checked my code and corrected any misspellings

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. *Any change needs to be discussed before proceeding.*

Fixes # (issue)

Issue #30 

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

A couple of fixes here:

(1) - Fixes the number of queries completed to be updated properly and actually trigger the delegate when appropriate and restructured the way errors were being concatenated. Added a custom field in the struct to tell you what index friend you are on for the current iteration.
(2) - Had to add in a constructor for use of friends in OnlineSubsystemEpicTypes.h that would be used to query their names. The GetBytes() method correctly sets the index to where we are filling the account id with the actual epic account id, as before it was using the product id - causing queries of names to fail miserably.

Now, in the log the callback result shows up and upon the callback in LogTemp, it now prints the name and spits out the right number of ids:

```
[2020.07.09-17.35.09:630][836]LogOnlineUser: Display: OSS: Query user info successful.
[2020.07.09-17.35.09:630][836]LogTemp: Query user callback result: 1 with number of ids: 1
[2020.07.09-17.35.09:630][836]LogOnlineUser: OSS: User name is: MPLEX VR
```

This is going to go hand-in-hand with my FriendsSubsystem implementation if this is integrated as that is where I could test more queries easily.

### CODE REVIEW

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp [hunk 1]
```diff
@@ -1,9 +1,11 @@
 #include "OnlinePrensenceEpic.h"
 #include "eos_presence.h"
 #include "OnlineSubsystemEpicTypes.h"
+
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp:4]: Please don't add additional whitespace and newlines to existing code.

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp [hunk 2]
```diff
@@ -478,8 +481,7 @@ EOnlineCachedResult::Type FOnlinePresenceEpic::GetCachedPresence(const FUniqueNe
 				// Get the last time the querying user was online.
 				FString lastOnlineString;
 				userAcc->GetUserAttribute(USER_ATTR_LAST_LOGIN_TIME, lastOnlineString);
-				OutPresence->LastOnline = FDateTime::FromUnixTimestamp(FCString::Atoi64(*lastOnlineString));
-
+
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp:484]: Why was the last online time removed? It's never added anywhere else.

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.h [hunk 3]
```diff
@@ -36,3 +36,5 @@ class FOnlinePresenceEpic
 
 	virtual EOnlineCachedResult::Type GetCachedPresenceForApp(const FUniqueNetId& LocalUserId, const FUniqueNetId& User, const FString& AppId, TSharedPtr<FOnlineUserPresence>& OutPresence) override;
 };
+
+typedef TSharedPtr<class FOnlinePresenceEpic, ESPMode::ThreadSafe> FOnlinePresenceEpicPtr;
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.h:40]: Please add a newline to the end of files. This is needed for UNIX compatability.

File: Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpic.cpp [hunk 4]
```diff
@@ -382,8 +382,8 @@ bool FOnlineSubsystemEpic::Shutdown()
 	DESTRUCT_INTERFACE(IdentityInterface);
 	DESTRUCT_INTERFACE(SessionInterface);
 	DESTRUCT_INTERFACE(UserInterface);
-	DESTRUCT_INTERFACE(this->PresenceInterface);
-
+	DESTRUCT_INTERFACE(PresenceInterface);
+
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpic.cpp:386]: The changes here were unecessary, so please remove the whitespace and re-add the removed `this->`.

File: Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpicTypes.h [hunk 5]
```diff
@@ -69,6 +69,7 @@ class FUniqueNetIdEpic
 		, epicAccountId(nullptr)
 	{
 	}
+
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpicTypes.h:72]: Same here, please remove superfluous newlines

File: Source/OnlineSubsystemEpic/Private/OnlineUserInterfaceEpic.cpp [hunk 6]
```diff
@@ -138,22 +141,33 @@ void FOnlineUserEpic::OnEOSQueryUserInfoComplete(EOS_UserInfo_QueryUserInfoCallb
 	thisPtr->UserQueryLock.Lock();
 
 	// Auto used below to increase readability
-	auto query = thisPtr->userQueries.Find(additionalData->StartTime);
-
+	TTuple<TArray<TSharedRef<FUniqueNetId const>>, TArray<bool>, TArray<FString>>* query = thisPtr->userQueries.Find(additionalData->StartTime);
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlineUserInterfaceEpic.cpp:144]: `auto` was used for a reason. Please restore it for better readability.

File: Source/OnlineSubsystemEpic/Private/OnlineUserInterfaceEpic.cpp [hunk 7]
```diff
@@ -710,6 +729,9 @@ TSharedPtr<FOnlineUser> FOnlineUserEpic::GetUserInfo(int32 LocalUserNum, const c
 		{
 			FUniqueNetIdEpic const epicUserId = static_cast<FUniqueNetIdEpic const>(UserId);
 
+			UE_LOG_ONLINE_USER(Log, TEXT("Local: %s:%s"), __FUNCTIONW__, *localUserId->ToDebugString());
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlineUserInterfaceEpic.cpp:732]: `__FUNCTIONW__` is a nonstandard addition to the compiler by Microsoft, use `__FUNCTION__` instead

File: Source/OnlineSubsystemEpic/Private/OnlineUserInterfaceEpic.h [hunk 8]
```diff
@@ -25,7 +25,7 @@ class FOnlineUserEpic
 
 	/** A list of all user ids for which the SDK has cached user information. */
 	TArray<EOS_EpicAccountId> queriedUserIdsCache;
-
+
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlineUserInterfaceEpic.h:28]: Again: Unnecessary whitespace

File: Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpicTypes.h [hunk 9]
```diff
@@ -318,6 +327,22 @@ class FUniqueNetIdEpic
 	}
 };
 
+/**
+ * An enumeration of the different friendship statuses. Modified from eos_friends_types.h
+ */
+UENUM()
+enum class EFriendStatus : uint8
```
Comment (Ruhrpottpatriot) [inline Source/OnlineSubsystemEpic/Private/OnlineSubsystemEpicTypes.h:334]: You don't need to explicitly mark the enum values with integers. This is done by the compiler.

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp (hunk 1 above)
Comment (MikhailSorokin) [reply Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp:4]: Yea, my b I'll be more cautious about that going forward

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp (hunk 2 above)
Comment (MikhailSorokin) [reply Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp:484]: LastOnline is not a field that exists in the FOnlinePresence struct. I'll put it back in but comment it out and explain what I just said there.

**Comment by MikhailSorokin** [comment]
Done, thanks for the pointers!

As a side note, please change the spelling errors in the Presence interface at some point. "OnlinePrensenceInterface" drives me crazy lol

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp (hunk 2 above)
Comment (Ruhrpottpatriot) [reply Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp]: Actually it is. In OnlinePresenceInterface.h L184 `LastOnline` is declared.

**Comment by Ruhrpottpatriot** [comment]
Copying this from our Discord conversation:
> also since I have done presence and friends succesfully in the EOS plugin project, I think I can handle the errors that are happening in the presence interface
there are some things you are assuming that aren't quite right in the presence interface - starting with the notification listener, which should be for every new user that has its presence queried
so a map of the user's account id to a notification handle is needed, in which one could unsubscribe for that user later
that is how it is done in the sample SDK as well

Yeah, if you could do that, that'd be great. For the friends stuff, can you open a new PR, or are the Presence fixes and Friends implementation too intertwined?

File: Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp (hunk 2 above)
Comment (MikhailSorokin) [reply Source/OnlineSubsystemEpic/Private/OnlinePrensenceEpic.cpp]: Ah, I see where the confusion is. It is _not_ a field in 4.24.3, but is in 4.25. It doesn't seem to be used right now though, can that just be removed for now though? The only place it seems to be used is in SocialUser in the Party class and that doesn't seem that it will be implemented for a while.
